# Multi Class - predict single category from multiple categories for a  data point

- Logistic Regression with multi_class ( soft_max ) 
- OneVsRest Logistic Regression with a BASE binary classifer
    can be use multiple labels if y is a 2D matrix of **0s or 1s** **only**
- OneVsOne Logistic Regression with a BASE binary classifier

# Multi Label - predict multiple categories from multiple categories for a  data point

- MultiOutputClassifier with a BASE binary/multi class classifier

In [1]:
import numpy as np
import pandas as pd

from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

In [2]:
# Note that lr takes string labels as well
test_lr=LogisticRegression(multi_class='multinomial',solver='lbfgs')
test_lr.fit([[1,2,3,4],[4,3,2,1]],['P','Q'])
test_lr.predict([[4,2,3,3]])
confusion_matrix(['P'],test_lr.predict([[4,2,3,3]]))

array([[0, 1],
       [0, 0]], dtype=int64)

In [3]:
iris=load_iris()
print(iris.data.shape,iris.target.shape)
# The prediction was too accurate . SO i have made training_size << test_size
split= train_test_split(iris.data,iris.target,test_size=0.75,random_state=25)
A= split[0]
B=split[2]
C=split[1]
D=split[3]

(150, 4) (150,)


In [4]:
lr=LogisticRegression(multi_class='multinomial',solver='lbfgs')
#lr.fit(iris.data,iris.target)
lr.fit(A,B)

LogisticRegression(multi_class='multinomial')

In [5]:
iris.data[-2:,:].shape,iris.data[-1:,:].shape,iris.data[-1,:].shape, iris.target[-1].shape,iris.target[-1:].shape
#iris.data[-1:,:]


((2, 4), (1, 4), (4,), (), (1,))

In [6]:
iris.target[-100:]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [7]:
#pred_classes=lr.predict(iris.data[-100:,:])
#pred_probs=lr.predict_proba(iris.data[-100:,:])

In [8]:
#predict takes 2D Array !!!!
pred_classes=lr.predict(C)
pred_probs=lr.predict_proba(C)

print(pred_classes.shape,pred_probs.shape)
print(pred_probs)

(113,) (113, 3)
[[9.24828716e-01 7.51528698e-02 1.84139895e-05]
 [5.14288601e-03 4.81816483e-01 5.13040631e-01]
 [9.74730444e-03 3.77111350e-01 6.13141346e-01]
 [2.85214537e-02 5.82142495e-01 3.89336052e-01]
 [3.38176294e-04 8.94896073e-02 9.10172216e-01]
 [2.33456851e-02 7.34413664e-01 2.42240651e-01]
 [5.85345072e-03 3.50364333e-01 6.43782216e-01]
 [9.48747996e-01 5.12394754e-02 1.25282403e-05]
 [9.16209975e-02 8.57147402e-01 5.12316003e-02]
 [1.83880419e-01 7.91627455e-01 2.44921257e-02]
 [9.31663349e-01 6.83077285e-02 2.89220215e-05]
 [8.98394804e-01 1.01566659e-01 3.85369767e-05]
 [9.04781258e-01 9.51698998e-02 4.88424492e-05]
 [3.94662870e-03 3.79936489e-01 6.16116882e-01]
 [9.42848195e-01 5.71327672e-02 1.90376275e-05]
 [1.49975338e-02 6.91916689e-01 2.93085777e-01]
 [1.65338561e-04 7.68992065e-02 9.22935455e-01]
 [6.99740444e-04 1.38723217e-01 8.60577042e-01]
 [1.32329245e-01 8.04655982e-01 6.30147731e-02]
 [4.34541218e-02 7.66439335e-01 1.90106543e-01]
 [1.28975501e-01 8.38248

In [9]:
def all_metrics(D,pred_classes):
    return (confusion_matrix(D,pred_classes) ,\
    recall_score(D,pred_classes,average=None) , precision_score(D,pred_classes,average=None), \
    f1_score(D,pred_classes,average=None) )

In [10]:
all_metrics(D,pred_classes)

(array([[41,  0,  0],
        [ 0, 35,  4],
        [ 0,  0, 33]], dtype=int64),
 array([1.       , 0.8974359, 1.       ]),
 array([1.        , 1.        , 0.89189189]),
 array([1.        , 0.94594595, 0.94285714]))

In [17]:
pred_classes.shape

(113,)

### Other Multi Class Classification

## MultiOutput ( predict multiple categories from multipe categories for a  data point)
- generates num_of_models = num_of_labels
- Each column is treated an independent/separate model 
- ** Target is 2D **
- Target is generally 
- For each data point , target is an array of num_of_labels [0 ,1, 1]
- For regular Classification, for each data point is target is a scalar of the label - 2
- each model = one for each column/label 
- so probabilities array shape = num_of_models shape
- so classes array shape = num_of_test_data_points * num_labels
- each output picks the label in each model  [1 1 0] for example


In [11]:
from sklearn.datasets import make_multilabel_classification
from sklearn.multioutput import MultiOutputClassifier

X, y = make_multilabel_classification(n_classes=3, random_state=0)
X.shape , X ,y


((100, 20),
 array([[3., 6., 1., ..., 1., 5., 0.],
        [3., 5., 5., ..., 1., 1., 1.],
        [3., 3., 5., ..., 0., 2., 1.],
        ...,
        [3., 7., 3., ..., 2., 4., 2.],
        [7., 2., 1., ..., 1., 1., 1.],
        [3., 5., 3., ..., 1., 2., 4.]]),
 array([[0, 1, 0],
        [0, 1, 0],
        [1, 1, 1],
        [1, 1, 1],
        [0, 1, 0],
        [1, 1, 0],
        [0, 0, 0],
        [1, 0, 0],
        [0, 0, 1],
        [0, 0, 0],
        [0, 1, 0],
        [0, 0, 0],
        [0, 1, 1],
        [1, 0, 0],
        [1, 1, 0],
        [1, 1, 1],
        [0, 0, 0],
        [0, 1, 0],
        [0, 0, 0],
        [1, 1, 1],
        [0, 1, 1],
        [0, 0, 1],
        [1, 0, 1],
        [0, 1, 0],
        [0, 1, 0],
        [1, 1, 0],
        [0, 1, 1],
        [0, 0, 1],
        [1, 1, 1],
        [1, 1, 0],
        [0, 1, 0],
        [0, 0, 1],
        [0, 0, 1],
        [0, 1, 0],
        [0, 0, 0],
        [0, 1, 0],
        [0, 1, 0],
        [1, 1, 1],
        [1, 1, 1]

In [19]:
clf = MultiOutputClassifier(LogisticRegression()).fit(X, y)
m_pred_classes=clf.predict(X[-2:])
m_pred_probs=clf.predict_proba(X[-2:])

print(m_pred_classes,m_pred_classes.shape,m_pred_probs,len(m_pred_probs))
#m_pred_probs.shape


[[1 1 1]
 [1 0 1]] (2, 3) [array([[0.14044346, 0.85955654],
       [0.16267135, 0.83732865]]), array([[0.09446415, 0.90553585],
       [0.78465578, 0.21534422]]), array([[0.453594  , 0.546406  ],
       [0.38098685, 0.61901315]])] 3


Model 1 - 2 datapoints X binary 
[array([[0.14044346, 0.85955654],
       [0.16267135, 0.83732865]]), 
Model 2 - 2 datapoints X binary       
array([[0.09446415, 0.90553585],
       [0.78465578, 0.21534422]]), 
Model 3 - 2 datapoints X binary        
 array([[0.453594  , 0.546406  ],
       [0.38098685, 0.61901315]])] 3

### if y contains only single column only one model is generated
- here y is a single column with (0,1,2) as class labels
- so predication is an array with single column with 0 or 1 or 2

In [13]:
z=np.zeros((100,1))
z[-25:]=2
z[25:50]=1
z

clf = MultiOutputClassifier(LogisticRegression(max_iter=500)).fit(X, z)

m_pred_classes=clf.predict(X[-10:])
m_pred_probs=clf.predict_proba(X[-10:])

m_pred_classes.shape,m_pred_classes,m_pred_probs
#,m_pred_probs.shape

((10, 1),
 array([[2.],
        [2.],
        [2.],
        [2.],
        [1.],
        [0.],
        [2.],
        [2.],
        [0.],
        [1.]]),
 [array([[0.29735607, 0.01669833, 0.6859456 ],
         [0.35424857, 0.03472184, 0.61102959],
         [0.28626994, 0.08585234, 0.62787772],
         [0.38757585, 0.01791295, 0.59451121],
         [0.3591262 , 0.5287604 , 0.11211341],
         [0.54803632, 0.08543395, 0.36652973],
         [0.13241586, 0.01890283, 0.84868131],
         [0.34492806, 0.19627413, 0.45879782],
         [0.51556317, 0.08515241, 0.39928442],
         [0.4569614 , 0.46048421, 0.08255439]])])

# Multi Class - predict single category from multiple categories for a  data point
## OneVsRestClassifier
- generates num_of_models = num_of_labels
- each model = one vs rest
- so probabilities array shape = num_of_models shape 
(technically each model generates 2/binary probabilities , Yes Prob =0... No Prob =0...)
For this classifier , Yes Prob from each model is taken and highest Yes Prob decides the class
- so classes array shape = num_of_test_data_points
- one class with highest probability among all probabilities is picked
- Note: if probs are 0.2 ,0.4, 0.3 the second one is picked. it need not be >0.5 cutoff like in softmax logistic

In [21]:

from sklearn.multiclass import OneVsRestClassifier

ovr_x = np.array([
    [10, 10],
    [8, 10],
    [-5, 5.5],
    [-5.4, 5.5],
    [-20, -20],
   [-15, -20]
])
over_y = np.array([0, 0, 1, 1, 2, 2])
ovr = OneVsRestClassifier(LogisticRegression()).fit(ovr_x, over_y)
ovr_preds_classes=ovr.predict([[-19, -20],[9,9]])
ovr_preds_probs=ovr.predict_proba([[-19, -20],[9,9]])

ovr_preds_classes,ovr_preds_classes.shape,ovr_preds_probs,ovr_preds_probs.shape


(array([2, 0]),
 (2,),
 array([[2.13656050e-07, 1.98122741e-02, 9.80187512e-01],
        [9.79177154e-01, 2.05881382e-02, 2.34708228e-04]]),
 (2, 3))

## OneVs One

- generates num_of_models = num_of_labels Combination 2 ( nC2 )
- each model = pair of class labels
- so probabilities array shape = num_of_models shape
- so classes array shape = num_of_test_data_points
- one class with highest probability among all probabilities is picked
- Note: if probs are 0.2 ,0.4, 0.3 the second one is picked. it need not be >0.5 cutoff like in softmax logistic

https://machinelearningmastery.com/one-vs-rest-and-one-vs-one-for-multi-class-classification/
Each binary classification model may predict one class label and the model with the most predictions or votes is predicted by the one-vs-one strategy.


An alternative is to introduce K(K − 1)/2 binary discriminant functions, one for every possible pair of classes. This is known as a one-versus-one classifier. Each point is then classified according to a majority vote amongst the discriminant functions.

In [27]:
from sklearn.multiclass import OneVsOneClassifier

ovo_x = np.array([
    [10, 10],
    [8, 10],
    [-5, 5.5],
    [-5.4, 5.5],
    [-20, -20],
   [-15, -20]
])
## 4 classes => 6 models
ovo_y = np.array([0, 0, 1, 1, 2, 3])
ovo = OneVsOneClassifier(LogisticRegression()).fit(ovo_x, ovo_y)
ovo_preds_classes=ovo.predict([[-19, -20],[9,9]])

## HAS NO predict_proba metho
#ovo_preds_probs=ovo.predict_proba([[-19, -20],[9,9]])

ovo_preds_classes,ovo_preds_classes.shape
#ovo_preds_probs,ovo_preds_probs.shape

(array([2, 0]), (2,))

In [31]:
from sklearn.multiclass import OneVsRestClassifier

ovr2_x = np.array([
    [10, 10],
    [8, 10],
    [-5, 5.5],
    [-5.4, 5.5],
    [-20, -20],
   [-15, -20]
])
## Multioutput target data is not supported with label binarization
##over2_y = np.array([[0,1],[0,2],[1,0],[1,1], [2,1], [2,1]])
## It has to be BINARY labels
over2_y = np.array([[0,1],[0,1],[1,0],[1,1], [1,1], [0,1]])
ovr2 = OneVsRestClassifier(LogisticRegression()).fit(ovr2_x, over2_y)

ovr2_preds_classes=ovr2.predict([[-19, -20],[9,9]])
ovr2_preds_probs=ovr2.predict_proba([[-19, -20],[9,9]])

ovr2_preds_classes,ovr2_preds_classes.shape,ovr2_preds_probs,ovr2_preds_probs.shape


(array([[1, 1],
        [0, 1]]),
 (2, 2),
 array([[0.77027017, 0.98508217],
        [0.00621082, 0.97810504]]),
 (2, 2))